# Biostat M280 Homework 3

**Due Friday, May 25 @ 11:59PM**

## Q1 - Big $n$ regression

Those who took my _203B: Introduction to Data Science_ last quarter had a (painful) experience of wrangling an Apache Spark cluster to do linear regression on a dataset with more than 100 million observations. Now we learnt various methods for solving linear regression and should realize that, with right choice of algorithm, it is a problem that can be handled by any moderate computer.

### Q1(1)

Download the flight data from <http://stat-computing.org/dataexpo/2009/the-data.html>. For this exercise, we only need data from years 2003-2008. If you are using Mac or Linux, you can run the following Bash script, which downloads and unzips files for all years.
```bash
# Download flight data by year
for i in {1987..2008}
  do
    echo "$(date) $i Download"
    fnam=$i.csv.bz2
    wget -O ./$fnam http://stat-computing.org/dataexpo/2009/$fnam
    echo "$(date) $i unzip"
    bzip2 -d ./$fnam
  done

# Download airline carrier data
wget -O ./airlines.csv http://www.transtats.bts.gov/Download_Lookup.asp?Lookup=L_UNIQUE_CARRIERS

# Download airports data
wget -O ./airports.csv https://raw.githubusercontent.com/jpatokal/openflights/master/data/airports.dat
```
Find out how many data points in each year.

In [42]:
@show c2003 = countlines("2003.csv")
@show c2004 = countlines("2004.csv")
@show c2005 = countlines("2005.csv")
@show c2006 = countlines("2006.csv")
@show c2007 = countlines("2007.csv")
@show c2008 = countlines("2008.csv")
@show total_lines = c2003 + c2004 + c2005 + c2006 + c2007 + c2008

c2003 = countlines("2003.csv") = 6488541
c2004 = countlines("2004.csv") = 7129271
c2005 = countlines("2005.csv") = 7140597
c2006 = countlines("2006.csv") = 7141923
c2007 = countlines("2007.csv") = 7453216
c2008 = countlines("2008.csv") = 7009729
total_lines = c2003 + c2004 + c2005 + c2006 + c2007 + c2008 = 42363277


42363277

### Q1(2) 

We are interested in how the time gain of a flight, defined as `DepDelay - ArrDelay`, depends on the distance traveled (`Distance`), departure delay (`DepDelay`), and carrier (`UniqueCarrier`). 

We want to fit a linear regression `Gain ~ 1 + Distance + DepDelay + UniqueCarrier` using data from 2003-2008. Note `UniqueCarrier` is a factor with 23 levels: "9E", "AA", "AQ", "AS", "B6", "CO", "DH", "DL", "EV", "F9", "FL", "HA", "HP", "MQ", "NW", "OH", "OO", "TZ", "UA", "US", "WN", "XE", and "YV". We use the dummy coding with "9E" as base level.

Will the design matrix (in double precision) fit into the memory of you computer?

**(2)** Data from 2003-2008 contain 42363277 observations. The design matrix has dimension $42363277 \times 26$, so in total $42363277 \times 26$ of double-precision numbers. Every double-precision number is 8 bytes. We need a memory of $42363277 \times 26 \times 8 / 10^9 \sim $8.81GB to fit the whole matrix. My poor MacBook with a 8GB memory can not handle it.

### Q1(3)

Review the [Summary of Linear Regression](http://hua-zhou.github.io/teaching/biostatm280-2018spring/slides/12-linreg/linreg.html) and devise a strategy to solve the linear regression.

Report the estimated regression coefficients $\widehat \beta$, estimated variance $\widehat \sigma^2 = \sum_i (y_i - \widehat y_i)^2 / (n - 1)$, and coefficient standard errors.

Hint: It took my laptop less than 3 minutes to import data and fit linear regression.

In [46]:
using SweepOperator

srand(280)

X = randn(5, 3) # predictor matrix
y = randn(5)    # response vector

# form the augmented Gram matrix
G = [X y]' * [X y]

4×4 Array{Float64,2}:
  9.589    -4.98208  -5.70052  -1.83933
 -4.98208   4.94476   3.81663   2.82462
 -5.70052   3.81663   5.45442   2.46008
 -1.83933   2.82462   2.46008   4.98343

In [8]:
# mapping from variable names to X columns
# carrier "9E" is used as base level
const var2col = Dict(
        "Intercept" => 1,
        "Distance" => 2,
        "DepDelay" => 3,
        "AA" => 4,
        "AQ" => 5,
        "AS" => 6,
        "B6" => 7,
        "CO" => 8,
        "DH" => 9,
        "DL" => 10,
        "EV" => 11,
        "F9" => 12,
        "FL" => 13,
        "HA" => 14,
        "HP" => 15,
        "MQ" => 16,
        "NW" => 17,
        "OH" => 18,
        "OO" => 19,
        "TZ" => 20,
        "UA" => 21,
        "US" => 22,
        "WN" => 23,
        "XE" => 24,
        "YV" => 25,
        "Gain" => 26)
# mapping from column to variable names
const col2var = map(reverse, var2col)

# a custom function to generate [X y] from data table
function generate_xy(tbl::NextTable)
    # X matrix
    XY = zeros(length(tbl), 26)
    # intercept term
    @views fill!(XY[:, 1], 1)
    # Distance term
    @views copy!(XY[:, 2], columns(tbl, :Distance))
    # DepDelay term
    @views copy!(XY[:, 3], columns(tbl, :DepDelay))
    # Dummy coding for airline
    @inbounds for i in 1:length(tbl)
        tbl[i][:UniqueCarrier] == "9E" && continue # base level
        XY[i, var2col[tbl[i][:UniqueCarrier]]] = 1
    end
    # last column is response: gain = depdelay - arrdelay
    XY[:, 26] = select(tbl, 
        (:DepDelay, :ArrDelay) => p -> Float64(p.DepDelay - p.ArrDelay))
    # return
    XY
end

generate_xy (generic function with 1 method)

In [1]:
using JuliaDB
# only need columns: DepDelay, ArrDelay, UniqueCarrier, Distance
@time yrtable2003 = loadtable(
    "2003.csv", 
    datacols = ["DepDelay", "ArrDelay", "UniqueCarrier", "Distance"])
@time yrtable2004 = loadtable(
    "2004.csv", 
    datacols = ["DepDelay", "ArrDelay", "UniqueCarrier", "Distance"])

 48.905613 seconds (153.54 M allocations: 8.160 GiB, 25.52% gc time)
 34.437851 seconds (31.53 M allocations: 4.276 GiB, 6.43% gc time)


Table with 7129270 rows, 4 columns:
DepDelay  ArrDelay  UniqueCarrier  Distance
───────────────────────────────────────────
-7        -14       "UA"           599
-9        -4        "UA"           599
3         5         "UA"           599
-3        -16       "UA"           599
5         3         "UA"           599
-2        -10       "UA"           599
20        29        "UA"           599
-3        -11       "UA"           599
-7        -12       "UA"           599
-4        -14       "UA"           599
1         8         "UA"           599
-2        -14       "UA"           599
⋮
0         -4        "DL"           821
-2        -9        "DL"           590
-2        -10       "DL"           532
-1        6         "DL"           391
2         -15       "DL"           599
-6        -15       "DL"           215
6         -1        "DL"           425
6         -1        "DL"           1541
6         3         "DL"           547
-2        -3        "DL"           545
-2        -13  

In [14]:
using SweepOperator
yrtable2003 = dropna(yrtable2003)
yrtable2004 = dropna(yrtable2004)
g3 = generate_xy(yrtable2003)
g4 = generate_xy(yrtable2004)
g = g3'*g3 
g

26×26 Array{Float64,2}:
      6.37569e6  4.5497e9         3.32999e7   …  0.0        1.03685e7 
      4.5497e9   5.28186e12       2.37964e10     0.0        1.01318e10
      3.32999e7  2.37964e10       4.51178e9      0.0        4.30844e7 
 738567.0        7.7143e8         3.42572e6      0.0        1.05689e6 
      0.0        0.0              0.0            0.0        0.0       
 157602.0        1.31742e8   895388.0         …  0.0   394580.0       
  66732.0        8.49327e7   377211.0            0.0   311411.0       
 299843.0        3.24764e8        1.1675e6       0.0    33212.0       
 280010.0        1.00559e8        2.63428e6      0.0   804746.0       
 652602.0        5.2721e8         2.95877e6      0.0   442798.0       
 266760.0        1.08515e8        2.37215e6   …  0.0   595560.0       
      0.0        0.0              0.0            0.0        0.0       
 142817.0        8.51543e7   983078.0            0.0  -211739.0       
   7792.0        4.77305e6     8050.0            0.0 

In [13]:
sweep!(g, 1:25)

26×26 Array{Float64,2}:
    NaN          NaN           …  NaN              NaN    NaN    NaN
      4.5497e9   NaN              NaN              NaN    NaN    NaN
      3.32999e7    2.37964e10     NaN              NaN    NaN    NaN
 738567.0          7.7143e8       NaN              NaN    NaN    NaN
      0.0          0.0            NaN              NaN    NaN    NaN
 157602.0          1.31742e8   …  NaN              NaN    NaN    NaN
  66732.0          8.49327e7      NaN              NaN    NaN    NaN
 299843.0          3.24764e8      NaN              NaN    NaN    NaN
 280010.0          1.00559e8      NaN              NaN    NaN    NaN
 652602.0          5.2721e8       NaN              NaN    NaN    NaN
 266760.0          1.08515e8   …  NaN              NaN    NaN    NaN
      0.0          0.0            NaN              NaN    NaN    NaN
 142817.0          8.51543e7      NaN              NaN    NaN    NaN
   7792.0          4.77305e6      NaN              NaN    NaN    NaN
 187266.0 

In [ ]:
# maybe a function
using JuliaDB
# pre-allocate a matrix gram_augmented  
n = 0
p = 26

gram_augmented = zeros(p, p)
gram_augmented_sum = zeros(p, p)
@time for year = 2003:2008
    filename = join([string(year),".csv"]) 
    yrtable = loadtable(
        file, 
        datacols = ["DepDelay", "ArrDelay", "UniqueCarrier", "Distance"])
    yrtable = dropna(yrtable)
    xy_matrix = generate_xy(yrtable)
    n += size()
    gram_augmented = At_mul_B(xy_matrix, xy_matrix)
    gram_augmented_sum += gram_augmented
end

In [ ]:
rand()

In [35]:
using SweepOperator

sweep!(gram_augmented_sum, 1:(p - 1))

26×26 Array{Float64,2}:
    NaN          NaN           …  NaN              NaN    NaN    NaN
      4.5497e9   NaN              NaN              NaN    NaN    NaN
      3.32999e7    2.37964e10     NaN              NaN    NaN    NaN
 738567.0          7.7143e8       NaN              NaN    NaN    NaN
      0.0          0.0            NaN              NaN    NaN    NaN
 157602.0          1.31742e8   …  NaN              NaN    NaN    NaN
  66732.0          8.49327e7      NaN              NaN    NaN    NaN
 299843.0          3.24764e8      NaN              NaN    NaN    NaN
 280010.0          1.00559e8      NaN              NaN    NaN    NaN
 652602.0          5.2721e8       NaN              NaN    NaN    NaN
 266760.0          1.08515e8   …  NaN              NaN    NaN    NaN
      0.0          0.0            NaN              NaN    NaN    NaN
 142817.0          8.51543e7      NaN              NaN    NaN    NaN
   7792.0          4.77305e6      NaN              NaN    NaN    NaN
 187266.0 

In [32]:
gram_augmented_sum

26×26 Array{Float64,2}:
    NaN           NaN             NaN          …  NaN          NaN    NaN
      2.72982e10  NaN             NaN             NaN          NaN    NaN
      1.99799e8     1.42778e11    NaN             NaN          NaN    NaN
      4.4314e6      4.62858e9       2.05543e7     NaN          NaN    NaN
      0.0           0.0             0.0           NaN          NaN    NaN
 945612.0           7.9045e8        5.37233e6  …  NaN          NaN    NaN
 400392.0           5.09596e8       2.26327e6     NaN          NaN    NaN
      1.79906e6     1.94858e9       7.005e6       NaN          NaN    NaN
      1.68006e6     6.03353e8       1.58057e7     NaN          NaN    NaN
      3.91561e6     3.16326e9       1.77526e7     NaN          NaN    NaN
      1.60056e6     6.51089e8       1.42329e7  …  NaN          NaN    NaN
      0.0           0.0             0.0           NaN          NaN    NaN
 856902.0           5.10926e8       5.89847e6     NaN          NaN    NaN
  46752.0     

In [25]:
  @time yrtable = loadtable(
        file, 
        datacols = ["DepDelay", "ArrDelay", "UniqueCarrier", "Distance"])

 35.663969 seconds (144.01 M allocations: 7.497 GiB, 34.53% gc time)


Table with 6488540 rows, 4 columns:
DepDelay  ArrDelay  UniqueCarrier  Distance
───────────────────────────────────────────
-4        -1        "UA"           837
-1        -3        "UA"           837
29        23        "UA"           837
-2        -9        "UA"           1835
18        52        "UA"           1835
-4        6         "UA"           1835
-4        -8        "UA"           1835
0         2         "UA"           1835
-4        19        "UA"           1835
3         4         "UA"           413
-4        -23       "UA"           413
-3        -19       "UA"           413
⋮
#NA       #NA       "DL"           1891
29        62        "DL"           581
39        66        "DL"           1891
26        27        "DL"           1678
114       134       "DL"           946
44        53        "DL"           813
16        47        "DL"           432
50        54        "DL"           432
-3        -5        "DL"           453
3         3         "DL"           689
-1     

In [27]:
using CSV
@time CSV.read("2003.csv")

LoadError: CSV.CSVError("error parsing a `Int64` value on column 13, row 5331553; encountered 'N'")

In [ ]:
using JuliaDB
# pre-allocate a matrix gram_augmented  
p = 26

gram_augmented = zeros(p, p)
gram_augmented_sum = zeros(p, p)
@time for year = 2003:2004
    filename = join([string(year),".csv"]) 
    yrtable = loadtable(
        file, 
        datacols = ["DepDelay", "ArrDelay", "UniqueCarrier", "Distance"])
    yrtable = dropna(yrtable)
    xy_matrix = generate_xy(yrtable)
    At_mul_B!(gram_augmented, xy_matrix, xy_matrix)
    gram_augmented_sum .= gram_augmented_sum + gram_augmented
end

In [ ]:
sweep!(gram_augmented_sum, 1:p - 1)

In [23]:
x = rand(5000, 5000);
y = rand(5000, 5000);
@time x .= x + y;
@time x = x + y;
@time x .= x .+ y;

  0.167099 seconds (6 allocations: 190.735 MiB, 5.77% gc time)
  0.199209 seconds (6 allocations: 190.735 MiB, 19.32% gc time)
  0.319038 seconds (31 allocations: 190.736 MiB, 57.87% gc time)


In [18]:
xy_matrix = generate_xy(yrtable)
gram_augmented = zeros(p, p)
At_mul_B!(gram_augmented, xy_matrix, xy_matrix)

26×26 Array{Float64,2}:
      6.37569e6  4.5497e9         3.32999e7   …  0.0        1.03685e7 
      4.5497e9   5.28186e12       2.37964e10     0.0        1.01318e10
      3.32999e7  2.37964e10       4.51178e9      0.0        4.30844e7 
 738567.0        7.7143e8         3.42572e6      0.0        1.05689e6 
      0.0        0.0              0.0            0.0        0.0       
 157602.0        1.31742e8   895388.0         …  0.0   394580.0       
  66732.0        8.49327e7   377211.0            0.0   311411.0       
 299843.0        3.24764e8        1.1675e6       0.0    33212.0       
 280010.0        1.00559e8        2.63428e6      0.0   804746.0       
 652602.0        5.2721e8         2.95877e6      0.0   442798.0       
 266760.0        1.08515e8        2.37215e6   …  0.0   595560.0       
      0.0        0.0              0.0            0.0        0.0       
 142817.0        8.51543e7   983078.0            0.0  -211739.0       
   7792.0        4.77305e6     8050.0            0.0 

### Q1(4)

Go to your resume/cv and claim you have experience performing analytics on data with hundred millions of observations.

### Sample code

Following code explores the data in 2003 and generates the design matrix and responses for that year. Feel free to use the code in your solution.

In [11]:
;head 2003.csv

Year,Month,DayofMonth,DayOfWeek,DepTime,CRSDepTime,ArrTime,CRSArrTime,UniqueCarrier,FlightNum,TailNum,ActualElapsedTime,CRSElapsedTime,AirTime,ArrDelay,DepDelay,Origin,Dest,Distance,TaxiIn,TaxiOut,Cancelled,CancellationCode,Diverted,CarrierDelay,WeatherDelay,NASDelay,SecurityDelay,LateAircraftDelay
2003,1,29,3,1651,1655,1912,1913,UA,1017,N202UA,141,138,119,-1,-4,ORD,MSY,837,5,17,0,NA,0,NA,NA,NA,NA,NA
2003,1,30,4,1654,1655,1910,1913,UA,1017,N311UA,136,138,108,-3,-1,ORD,MSY,837,2,26,0,NA,0,NA,NA,NA,NA,NA
2003,1,31,5,1724,1655,1936,1913,UA,1017,N317UA,132,138,110,23,29,ORD,MSY,837,5,17,0,NA,0,NA,NA,NA,NA,NA
2003,1,1,3,1033,1035,1625,1634,UA,1018,N409UA,232,239,215,-9,-2,OAK,ORD,1835,6,11,0,NA,0,NA,NA,NA,NA,NA
2003,1,2,4,1053,1035,1726,1634,UA,1018,N496UA,273,239,214,52,18,OAK,ORD,1835,13,46,0,NA,0,NA,NA,NA,NA,NA
2003,1,3,5,1031,1035,1640,1634,UA,1018,N412UA,249,239,223,6,-4,OAK,ORD,1835,13,13,0,NA,0,NA,NA,NA,NA,NA
2003,1,4,6,1031,1035,1626,1634,UA,1018,N455UA,235,239,219,-8,-4,OAK,ORD,183

In [2]:
# how many data points
countlines("2003.csv")

6488541

In [3]:
@time yrtable = loadtable(
    "2003.csv", 
    datacols = ["DepDelay", "ArrDelay", "UniqueCarrier", "Distance"])

 37.743440 seconds (144.03 M allocations: 7.497 GiB, 37.49% gc time)


Table with 6488540 rows, 4 columns:
DepDelay  ArrDelay  UniqueCarrier  Distance
───────────────────────────────────────────
-4        -1        "UA"           837
-1        -3        "UA"           837
29        23        "UA"           837
-2        -9        "UA"           1835
18        52        "UA"           1835
-4        6         "UA"           1835
-4        -8        "UA"           1835
0         2         "UA"           1835
-4        19        "UA"           1835
3         4         "UA"           413
-4        -23       "UA"           413
-3        -19       "UA"           413
⋮
#NA       #NA       "DL"           1891
29        62        "DL"           581
39        66        "DL"           1891
26        27        "DL"           1678
114       134       "DL"           946
44        53        "DL"           813
16        47        "DL"           432
50        54        "DL"           432
-3        -5        "DL"           453
3         3         "DL"           689
-1     

In [4]:
# drop rows with missing values
yrtable = dropna(yrtable)

Table with 6375689 rows, 4 columns:
DepDelay  ArrDelay  UniqueCarrier  Distance
───────────────────────────────────────────
-4        -1        "UA"           837
-1        -3        "UA"           837
29        23        "UA"           837
-2        -9        "UA"           1835
18        52        "UA"           1835
-4        6         "UA"           1835
-4        -8        "UA"           1835
0         2         "UA"           1835
-4        19        "UA"           1835
3         4         "UA"           413
-4        -23       "UA"           413
-3        -19       "UA"           413
⋮
70        66        "DL"           1891
29        62        "DL"           581
39        66        "DL"           1891
26        27        "DL"           1678
114       134       "DL"           946
44        53        "DL"           813
16        47        "DL"           432
50        54        "DL"           432
-3        -5        "DL"           453
3         3         "DL"           689
-1     

In [36]:
t = table([1,2],[3,4], names=[:x, :y])

Table with 2 rows, 2 columns:
x  y
────
1  3
2  4

In [6]:
@time xy = generate_xy(yrtable)

  3.000190 seconds (13.01 M allocations: 1.866 GiB, 32.33% gc time)


6375689×26 Array{Float64,2}:
 1.0   837.0   -4.0  0.0  0.0  0.0  …  0.0  1.0  0.0  0.0  0.0  0.0   -3.0
 1.0   837.0   -1.0  0.0  0.0  0.0     0.0  1.0  0.0  0.0  0.0  0.0    2.0
 1.0   837.0   29.0  0.0  0.0  0.0     0.0  1.0  0.0  0.0  0.0  0.0    6.0
 1.0  1835.0   -2.0  0.0  0.0  0.0     0.0  1.0  0.0  0.0  0.0  0.0    7.0
 1.0  1835.0   18.0  0.0  0.0  0.0     0.0  1.0  0.0  0.0  0.0  0.0  -34.0
 1.0  1835.0   -4.0  0.0  0.0  0.0  …  0.0  1.0  0.0  0.0  0.0  0.0  -10.0
 1.0  1835.0   -4.0  0.0  0.0  0.0     0.0  1.0  0.0  0.0  0.0  0.0    4.0
 1.0  1835.0    0.0  0.0  0.0  0.0     0.0  1.0  0.0  0.0  0.0  0.0   -2.0
 1.0  1835.0   -4.0  0.0  0.0  0.0     0.0  1.0  0.0  0.0  0.0  0.0  -23.0
 1.0   413.0    3.0  0.0  0.0  0.0     0.0  1.0  0.0  0.0  0.0  0.0   -1.0
 1.0   413.0   -4.0  0.0  0.0  0.0  …  0.0  1.0  0.0  0.0  0.0  0.0   19.0
 1.0   413.0   -3.0  0.0  0.0  0.0     0.0  1.0  0.0  0.0  0.0  0.0   16.0
 1.0   413.0    0.0  0.0  0.0  0.0     0.0  1.0  0.0  0.0  0.0  0.0   1

## Q2 - Google PageRank

We are going to try different numerical methods learnt in class on the [Google PageRank problem](https://en.wikipedia.org/wiki/PageRank).

### Q2(1)

Let $\mathbf{A} \in \{0,1\}^{n \times n}$ be the connectivity matrix of $n$ web pages with entries
$$
\begin{eqnarray*}
	a_{ij}= \begin{cases}
	1 & \text{if page $i$ links to page $j$} \\
	0 & \text{otherwise}
	\end{cases}.
\end{eqnarray*}
$$
$r_i = \sum_j a_{ij}$ is the out-degree of page $i$. That is $r_i$ is the number of links on page $i$. Imagine a random surfer exploring the space of $n$ pages according to the following rules.  

- From a page $i$ with $r_i>0$
    * with probability $p$, (s)he randomly chooses a link on page $i$ (uniformly) and follows that link to the next page  
    * with probability $1-p$, (s)he randomly chooses one page from the set of all $n$ pages (uniformly) and proceeds to that page  
- From a page $i$ with $r_i=0$ (a dangling page), (s)he randomly chooses one page from the set of all $n$ pages (uniformly) and proceeds to that page  
    
The process defines a Markov chain on the space of $n$ pages. Write the transition matrix $\mathbf{P}$ of the Markov chain as a sparse matrix plus rank 1 matrix.

### Q2(2)

According to standard Markov chain theory, the (random) position of the surfer converges to the stationary distribution $\mathbf{x} = (x_1,\ldots,x_n)^T$ of the Markov chain. $x_i$ has the natural interpretation of the proportion of times the surfer visits page $i$ in the long run. Therefore $\mathbf{x}$ serves as page ranks: a higher $x_i$ means page $i$ is more visited. It is well-known that $\mathbf{x}$ is the left eigenvector corresponding to the top eigenvalue 1 of the transition matrix $\mathbf{P}$. That is $\mathbf{P}^T \mathbf{x} = \mathbf{x}$. Therefore $\mathbf{x}$ can be solved as an eigen-problem. Show that it can also be cast as solving a linear system. Since the row sums of $\mathbf{P}$ are 1, $\mathbf{P}$ is rank deficient. We can replace the first equation by the $\sum_{i=1}^n x_i = 1$.

### Q2(3)

Download the [`ucla.zip`](http://hua-zhou.github.io/teaching/biostatm280-2018spring/hw/hw3/ucla.zip) package from course webpage. Unzip the package, which contains two files `U.txt` and `A.txt`. `U.txt` lists the 500 URL names. `A.txt` is the $500 \times 500$ connectivity matrix. Read data into Julia. Compute summary statistics:
* number of pages
* number of edges
* number of dangling nodes (pages with no out links)
* which page has max in-degree?
* which page has max out-degree?
* visualize the sparsity pattern of $\mathbf{A}$

### Q2(4)

Set the _teleportation_ parameter at $p = 0.85$. Try the following methods to solve for $\mathbf{x}$ using the `ucla.zip` data.

0. A dense linear system solver such as LU decomposition.  
0. A simple iterative linear system solver such as Jacobi or Gauss-Seidel.   
0. A dense eigen-solver.  
0. A simple iterative eigen-solver such as the power method.  

For iterative methods, you can use the [`IterativeSolvers.jl`](https://github.com/JuliaMath/IterativeSolvers.jl) package. Make sure to utilize the special structure of $\mathbf{P}$ (sparse + rank 1) to speed up the matrix-vector multiplication.

### Q2(5)

List the top 20 ranked URLs you found.

### Q2(6)

As of Monday May 11 2018, there are at least 1.83 billion indexed webpages on internet according to <http://www.worldwidewebsize.com/>. Explain whether each of these methods works for the PageRank problem at this scale.